File: 0000-0999/0354.RussianDollEnvelopes.py
class Solution(object):
    def maxEnvelopes(self, envelopes):
        """
        :type envelopes: List[List[int]]
        :rtype: int
        """
        if not envelopes:
            return 0
        envs = sorted(envelopes, key = lambda x: (x[0], -x[1]))
        end = [0] * len(envs)
        high = 0
        # O(nlogn)
        for w, h in envs:   # O(n)
            #i = bisect.bisect_left(end, h, 0, high) # O(logn)
            i = self.my_bisect_left(end, h, 0, high)    # O(logn)
            end[i] = h
            if i == high:
                high += 1
        return high
        
        ''' # O(n^2), TLE
        env = sorted(envelopes)
        dp = [1] * len(env)
        for i in xrange(1, len(env)):
            for j in xrange(i):
                if env[i][0] > env[j][0] and env[i][1] > env[j][1]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
        '''
        
    def my_bisect_left(self, nums, target, l, r):
        while l < r:
            m = l + (r-l)/2
            if target <= nums[m]:
                r = m
            else:
                l = m + 1
        return l
                

File: 0000-0999/0354.RussianDollEnvelopes.go
func maxEnvelopes(envelopes [][]int) int {
    sort.Slice(envelopes, func(i, j int) bool {
        if envelopes[i][0] == envelopes[j][0] {
            return envelopes[i][1] > envelopes[j][1]
        }
        return envelopes[i][0] < envelopes[j][0]
    })
    dp := make([]int, 0, len(envelopes))
    for _, e := range envelopes {
        idx := binInsertLeft(dp, e[1])
        if idx == len(dp) {
            dp = append(dp, e[1])
        } else {
            dp[idx] = e[1]
        }
    }
    return len(dp)
}

func binInsertLeft(nums []int, target int) int {
    l, r := 0, len(nums)
    for l < r {
        mid := (l+r) / 2
        if nums[mid] < target {
            l = mid + 1
        } else {
            r = mid
        }
    }
    return l
}